libs/libc/time/lib_localtime.c

```diff
@@ -2528,9 +2537,20 @@ void tzset(void)
 {
   FAR const char *name;
 
-  tz_semtake(&g_lcl_sem);
-
   name = getenv("TZ");
+  if (g_lcl_isset > 0 && name && strcmp(g_lcl_tzname, name) == 0)
+    {
+      return;
+    }
+
+#ifndef __KERNEL__
+  if (up_interrupt_context())
+    {
+      return;
+    }
```

Comment:
By the way, is this the correct place for performing this check?
Shouldn't it be checked in `nx_vsyslog()` before calling `localtime_r()`?

Reply:
If you "grep -r gmtime" in arch or drivers folder, more than 20 place call this function. If we allow kernel component call locale time related API, the check need be done here instead each call site.

Reply:
The point here is responsibility, which I think is important for having a well-defined interface.
Maybe we should add to time-related APIs that they should not be called from interrupt context.

Don't you think that, although an easier change, checking `if (up_interrupt_context())` here seems more like a quick fix?
For me it makes more sense to explicitly state the usage restriction on the APIs and fix the API users accordingly.

Reply:
Not being able to be called from interrupt context is a real limitation for time-related APIs.
There are similar examples throughout NuttX code where assumption and limitations are documented on function APIs:

https://github.com/apache/incubator-nuttx/blob/master/net/arp/arp.h#L253-L255

Reply:
I am fine with any approach, but the mainline will panic directly if you enable both config. And you can remove localtime from syslog by lose the local timing, but the call in driver is hard to remove.

Reply:
> If we allow kernel component call locale time related API, the check need be done here instead each call site.

I think I got your point. This restriction does not apply to drivers since this change only targets the non-kernel build.
Indeed, it would be a bit hard to check in each call site.

Reply:
So, Thanks @xiaoxiang781216 for detailed explanation and @gustavonihei for review. And i have deleted this redundant check for tzset. Thank you!